compiler/noirc_evaluator/src/ssa/ir/dfg/simplify/cast.rs

```diff
@@ -18,6 +18,10 @@ pub(super) fn simplify_cast(
 ) -> SimplifyResult {
     use SimplifyResult::*;
 
+    if Type::Numeric(dst_typ) == dfg.type_of_value(value) {
+        return SimplifiedTo(value);
+    }
```

Comment:
> but we're not simplifying out casts of constants into the same type. I've then boosted up the optimization for checking equality of the source and destination types so we always unwrap these

Do we only want to simplify numeric constants? It looks to me like `dfg.type_of_value(value)` could return the same type as `dst_type` if `value` is an `Instruction` that returns a number that isn't a constant.

It looks like the next `if let` is about unwrapping a 2-step cast into a 1-step cast, but if that wasn't the case, then the next `if let Some(constant)` only matched constants, except for the `else if` (which I don't have locally, it must be recent).

Reply:
Ah, ok, I guess the 2nd sentence means not just constants. Maybe the "known value" in the docstring is a bit vague without the check of being a constant, but removing a cast to the same type as the source makes sense.

Reply:
> Do we only want to simplify numeric constants? It looks to me like dfg.type_of_value(value) could return the same type as dst_type if value is an Instruction that returns a number that isn't a constant.

No, we want this optimization to apply on non-constant inputs as well as any conversion of a type to itself is a noop.

Reply:
We're basically just moving the else-block (which acted on non-constants) from the following if-statement up to the top of the function.